jupyterhub/app.py

```diff
@@ -2358,6 +2359,195 @@ async def init_api_tokens(self):
+            domain=domain,
+            host=host,
+            hub=self.hub,
+        )
+        traits = service.traits(input=True)
+        for key in traits:
+            value = orm_service.get_column(key)
+            if value is not None:
+                setattr(service, key, value)
+
+        if orm_service.oauth_client is not None:
+            service.oauth_redirect_uri = orm_service.oauth_client.redirect_uri
+            service.oauth_client_id = orm_service.oauth_client.identifier
+            service.oauth_redirect_uri = orm_service.oauth_client.redirect_uri
+
+        self._service_map[name] = service
+
+        return service
+
+    def service_from_spec(
+        self,
+        spec: Dict,
+        from_config=True,
+        domain: Optional[str] = None,
+        host: Optional[str] = None,
+    ) -> Optional[Service]:
+        """Create the service instance and related objects from
+        config data.
+
+        Args:
+            spec (Dict): The spec of service, defined in the config file.
+            from_config (bool, optional): `True` if the service will be created
+            from the config file, `False` if it is created from REST API.
+            Defaults to `True`.
+            domain (Optional[str]): The domain of Hub session. Defaults to None.
+            host (Optional[str]): The host of Hub session. Defaults to None.
+
+        Returns:
+            Optional[Service]: The created service
+        """
+
+        if domain is None:
+            if self.domain:
+                domain = 'services.' + self.domain
+            else:
+                domain = ''
+
+        if host is None:
+            if self.domain:
+                parsed = urlparse(self.subdomain_host)
+                host = f'{parsed.scheme}://services.{parsed.netloc}'
+            else:
+                host = ''
+
+        if 'name' not in spec:
+            raise ValueError('service spec must have a name: %r' % spec)
+
+        name = spec['name']
+        # get/create orm
+        orm_service = orm.Service.find(self.db, name=name)
+        if orm_service is None:
+            # not found, create a new one
+            orm_service = orm.Service(name=name, from_config=from_config)
+            self.db.add(orm_service)
+            if spec.get('admin', False):
+                self.log.warning(
+                    f"Service {name} sets `admin: True`, which is deprecated in JupyterHub 2.0."
+                    " You can assign now assign roles via `JupyterHub.load_roles` configuration."
+                    " If you specify services in the admin role configuration, "
+                    "the Service admin flag will be ignored."
+                )
+                roles.update_roles(self.db, entity=orm_service, roles=['admin'])
+        else:
+            # Do nothing if the config file tries to modify a API-base service
+            # or vice versa.
+            if orm_service.from_config != from_config:
+                return
```

Comment:
What's your thinking behind returning None instead of throwing an Exception? Should this log an self.log.error message?

Reply:
Throwing an exception here would prevent the startup of Jupyterhub. I added a log message to clarify the error. Returning `None` here helps the API handler to detect if the service is not created properly.

Reply:
I think it might be right for the Hub to halt. Invalid configuration errors like this should usually prevent loading the configuration. This method should not return without error if it did nothing, it should be up to the caller to catch that error, if it should at all.